### PR TITLE
 457033 - [IntelliJ] - Selection screen for fields

### PR DESCRIPTION
**Pull request by kdrtibor**

### CODE REVIEW

File: src/main/java/com/hpe/adm/octane/ideplugins/intellij/ui/detail/EntityDetailPresenter.java [hunk 1]
```diff
@@ -82,8 +86,13 @@ public void setEntity(Entity entityType, Long entityId) {
         RestUtil.runInBackground(
                 () -> {
                     try {
-                        entityModel = entityService.findEntity(this.entityType, this.entityId,metadataService.getFields(entityType));
-                        entityModel.setValue(new StringFieldModel("type",entityType.getSubtypeName()));
+                        Set<String> fields;
+                        if(entityType.isSubtype()) {
+                            fields	= new HashSet<>(metadataService.getFields(entityType.getSubtypeOf()));
```
Comment (atothhpe) [inline src/main/java/com/hpe/adm/octane/ideplugins/intellij/ui/detail/EntityDetailPresenter.java]: Weird spacing

File: src/main/java/com/hpe/adm/octane/ideplugins/intellij/ui/detail/EntityDetailPresenter.java [hunk 2]
```diff
@@ -93,7 +102,7 @@ public void setEntity(Entity entityType, Long entityId) {
                 (entityModel) -> {
                     if (entityModel != null) {
                         this.entityModel = entityModel;
-                        entityDetailView.setEntityModel(entityModel);
+                        entityDetailView.createDetailsPanel(entityModel,metadataService);
                         entityDetailView.setSaveSelectedPhaseButton(new SaveSelectedPhaseAction());
                         entityDetailView.setRefreshEntityButton(new EntityRefreshAction());
                         if (entityType!= TASK&&entityType != MANUAL_TEST_RUN && entityType != TEST_SUITE_RUN) {
```
Comment (atothhpe) [inline src/main/java/com/hpe/adm/octane/ideplugins/intellij/ui/detail/EntityDetailPresenter.java]: Weird spacing

File: src/main/java/com/hpe/adm/octane/ideplugins/intellij/ui/detail/EntityDetailPresenter.java [hunk 3]
```diff
@@ -118,7 +127,17 @@ public void setEntity(Entity entityType, Long entityId) {
...
 
     }
-
+    public void getUdfLabels(FormLayout octaneForms){
+        for(FormLayoutSection layoutSection: octaneForms.getFormLayoutSections()){
+            for(FormField fieldModel : layoutSection.getFields()){
+                if(fieldModel.getName().contains("_udf")){
+                    fieldModel.setName(metadataService.getUdfLabel(fieldModel.getName()));
+                    entityModel.getValue(fieldModel.getName());
+                    entityModel.setValue(new StringFieldModel(metadataService.getUdfLabel(fieldModel.getName()),entityModel.getValue(fieldModel.getName()).getValue().toString()));
```
Comment (atothhpe) [inline src/main/java/com/hpe/adm/octane/ideplugins/intellij/ui/detail/EntityDetailPresenter.java]: Weird formatting

File: src/main/java/com/hpe/adm/octane/ideplugins/intellij/ui/detail/EntityDetailPresenter.java [hunk 4]
```diff
@@ -109,16 +113,20 @@ public void setEntity(Entity entityType, Long entityId) {
...
                 },
                 null,
                 null,
                 "Loading entity " + entityType.name() + ": " + entityId);
+
+
     }
 
     private void setPossibleTransitions(EntityModel entityModel) {
+        Collection<EntityModel> result = new HashSet<>();
```
Comment (alexandra-marinescu) [inline src/main/java/com/hpe/adm/octane/ideplugins/intellij/ui/detail/EntityDetailPresenter.java]: Could have use Set here

File: src/main/java/com/hpe/adm/octane/ideplugins/intellij/ui/detail/EntityDetailPresenter.java [hunk 5]
```diff
@@ -134,12 +142,13 @@ private void setPossibleTransitions(EntityModel entityModel) {
...
 
     private void setComments(EntityModel entityModel) {
+        Collection<EntityModel> result = new HashSet<>();
         RestUtil.runInBackground(() -> commentService.getComments(entityModel), (comments) -> entityDetailView.setComments(comments), null, "Failed to get possible comments", "fetching comments");
     }
 
     private final class EntityRefreshAction extends AnAction {
         public EntityRefreshAction() {
-            super("Refresh current entity", "This will refresh the current entity.", IconLoader.findIcon(Constants.IMG_REFRESH_ICON));
+            super("Refresh current entity", "Refresh entity details", IconLoader.findIcon(Constants.IMG_REFRESH_ICON));
```
Comment (alexandra-marinescu) [inline src/main/java/com/hpe/adm/octane/ideplugins/intellij/ui/detail/EntityDetailPresenter.java]: why did you change the message and where is this message displayed?

File: src/main/java/com/hpe/adm/octane/ideplugins/intellij/ui/detail/EntityDetailPresenter.java [hunk 6]
```diff
@@ -150,17 +159,47 @@ public void actionPerformed(AnActionEvent e) {
...
     private final class EntityCommentsAction extends AnAction {
         public EntityCommentsAction() {
-            super("Show comments for current entity", "This will show comments for current entity.", IconLoader.findIcon(Constants.IMG_COMMENTS_ICON));
+            super("Show comments for current entity", "Show comments for current entity", IconLoader.findIcon(Constants.IMG_COMMENTS_ICON));
         }
 
         public void actionPerformed(AnActionEvent e) {
-            entityDetailView.getEntityDetailsPanel().activateCollapsible();
+            //GeneralEntityDetailsPanel.getCommetsDetails().getActionMap().get(JXCollapsiblePane.TOGGLE_ACTION);
+            //setComments(entityModel);
```
Comment (alexandra-marinescu) [inline src/main/java/com/hpe/adm/octane/ideplugins/intellij/ui/detail/EntityDetailPresenter.java]: remove comments

File: src/main/java/com/hpe/adm/octane/ideplugins/intellij/ui/detail/EntityDetailPresenter.java [hunk 7]
```diff
@@ -150,17 +159,47 @@ public void actionPerformed(AnActionEvent e) {
...
+    public final class SelectFieldsAction extends AnAction {
+
+        private boolean defaultfields = true;
+
+        public void setDefaultFieldsIcon(boolean defaultfields) {
+            this.defaultfields = defaultfields;
         }
+
+        public SelectFieldsAction() {
+            super("Select fields for this entity type", "Select fields popup.", IconLoader.findIcon(Constants.IMG_FIELD_SELECTION_DEFAULT));
```
Comment (alexandra-marinescu) [inline src/main/java/com/hpe/adm/octane/ideplugins/intellij/ui/detail/EntityDetailPresenter.java]: be consistent. why did you add . at the end?

File: src/main/java/com/hpe/adm/octane/ideplugins/intellij/ui/detail/EntityDetailPresenter.java [hunk 8]
```diff
@@ -150,17 +159,47 @@ public void actionPerformed(AnActionEvent e) {
...
+        }
+
+        public void actionPerformed(AnActionEvent e) {
+            entityDetailView.getEntityDetailsPanel().activateFieldsSettings();
+        }
+
+        public void update(AnActionEvent e){
+            if(defaultfields){
+                e.getPresentation().setIcon(IconLoader.findIcon(Constants.IMG_FIELD_SELECTION_DEFAULT));
```
Comment (alexandra-marinescu) [inline src/main/java/com/hpe/adm/octane/ideplugins/intellij/ui/detail/EntityDetailPresenter.java]: remove extra lines in if/else

File: src/main/java/com/hpe/adm/octane/ideplugins/intellij/ui/detail/FieldsSelectFrame.java [hunk 9]
```diff
@@ -0,0 +1,382 @@
...
+import java.awt.event.WindowEvent;
+import java.awt.event.WindowFocusListener;
+import java.util.*;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * Created by manasse on 11/8/2017.
+ */
+public class FieldsSelectFrame extends JFrame {
```
Comment (alexandra-marinescu) [inline src/main/java/com/hpe/adm/octane/ideplugins/intellij/ui/detail/FieldsSelectFrame.java]: add proper jdoc; make sure you don't have unused imports

File: src/main/java/com/hpe/adm/octane/ideplugins/intellij/ui/detail/FieldsSelectFrame.java [hunk 10]
```diff
@@ -0,0 +1,382 @@
...
+
+    private List<SelectionListener> listeners = new ArrayList<>();
+
+    public FieldsSelectFrame(Set<String> defaultFields, Collection<FieldMetadata> allFields, Map<Entity, Set<String>> selectedFieldsMap, Entity entityType, IdePluginPersistentState idePluginPersistentState, EntityDetailPresenter.SelectFieldsAction fieldsButton) {
+        this.defaultFields = defaultFields;
+        this.allFields = allFields;
+        this.selectedFieldsMap = selectedFieldsMap;
+        this.selectedFields = selectedFieldsMap.get(entityType);
+        this.idePluginPersistentState = idePluginPersistentState;
+        this.fieldsActionButton = fieldsButton;
```
Comment (alexandra-marinescu) [inline src/main/java/com/hpe/adm/octane/ideplugins/intellij/ui/detail/FieldsSelectFrame.java]: this constructor is huge. Split in methods if you actually need all the logic there

File: src/main/java/com/hpe/adm/octane/ideplugins/intellij/ui/detail/FieldsSelectFrame.java [hunk 11]
```diff
@@ -0,0 +1,382 @@
...
+            fieldsActionButton.setDefaultFieldsIcon(false);
+            resetButton.setEnabled(true);
+        }
+
+        setAlwaysOnTop(true);
+        setUndecorated(true);
+        pack();
+    }
+
+    public JScrollPane createFieldsPanel() {
```
Comment (alexandra-marinescu) [inline src/main/java/com/hpe/adm/octane/ideplugins/intellij/ui/detail/FieldsSelectFrame.java]: jdoc for public methods & fields

File: src/main/java/com/hpe/adm/octane/ideplugins/intellij/ui/detail/GeneralEntityDetailsPanel.java [hunk 12]
```diff
@@ -13,46 +13,87 @@
...
+
+
+    public GeneralEntityDetailsPanel(EntityModel entityModel, Collection<FieldMetadata> fields) {
         setLayout(new BorderLayout(0, 0));
 
+        this.entityModel = entityModel;
+        this.fields = fields;
+
+        DataManager dataManager = DataManager.getInstance();
+        @SuppressWarnings("deprecation")
```
Comment (alexandra-marinescu) [inline src/main/java/com/hpe/adm/octane/ideplugins/intellij/ui/detail/GeneralEntityDetailsPanel.java]: why do you need to use deprecated methods?

File: src/main/java/com/hpe/adm/octane/ideplugins/intellij/ui/detail/GeneralEntityDetailsPanel.java [hunk 13]
```diff
@@ -70,16 +111,37 @@ public GeneralEntityDetailsPanel(EntityModel entityModel) {
         gbc_headerPanel.gridy = 0;
         rootPanel.add(headerPanel, gbc_headerPanel);
 
-
-        entityDetailsPanel = drawSpecificDetailsForEntity(entityModel);
+        //create main details panel
```
Comment (alexandra-marinescu) [inline src/main/java/com/hpe/adm/octane/ideplugins/intellij/ui/detail/GeneralEntityDetailsPanel.java]: comment is useless; the method name says it all

File: src/main/java/com/hpe/adm/octane/ideplugins/intellij/ui/detail/GeneralEntityDetailsPanel.java [hunk 14]
```diff
@@ -179,286 +245,212 @@ public EntityModel getSelectedTransition() {
...
+        headerPanel.setFieldSelectButton(fieldSelectButton);
+        fieldsPopup = new FieldsSelectFrame(defaultFields.get(Entity.getEntityType(entityModel)),
+                fields,
+                selectedFields,
+                Entity.getEntityType(entityModel),
+                idePluginPersistentState,
+                fieldSelectButton);
+        fieldsPopup.addSelectionListener(e -> createSectionWithEntityDetails(entityModel, fieldsPopup.getSelectedFields()));
+        //listener for other opened tabs with the same entity
+        fieldsPopup.addSelectionListener(selectionListener);
```
Comment (alexandra-marinescu) [inline src/main/java/com/hpe/adm/octane/ideplugins/intellij/ui/detail/GeneralEntityDetailsPanel.java]: remove comment

File: src/main/java/com/hpe/adm/octane/ideplugins/intellij/ui/treetable/EntityTreeView.java [hunk 15]
```diff
@@ -165,6 +164,7 @@ public void mousePressed(MouseEvent e) {
                                 handler.entityDoubleClicked(e, entityType, entityId, entityModel);
                             } catch (Exception ex) {
                                 //TODO: logger and error bubble
+                                ex.printStackTrace();
```
Comment (alexandra-marinescu) [inline src/main/java/com/hpe/adm/octane/ideplugins/intellij/ui/treetable/EntityTreeView.java]: log the error. printing the stack trace does not help

File: src/main/java/com/hpe/adm/octane/ideplugins/intellij/ui/detail/EntityDetailPresenter.java (hunk 5 above)
Comment (kdrtibor) [reply src/main/java/com/hpe/adm/octane/ideplugins/intellij/ui/detail/EntityDetailPresenter.java]: I changed the message to be the same as the ones in eclipse. Common agreement with Madalina and Andras.

File: src/main/java/com/hpe/adm/octane/ideplugins/intellij/ui/detail/GeneralEntityDetailsPanel.java (hunk 12 above)
Comment (kdrtibor) [reply src/main/java/com/hpe/adm/octane/ideplugins/intellij/ui/detail/GeneralEntityDetailsPanel.java]: I need to use the deprecated method to obtain the intelliJ Project in order to use it to get the plugin persistent state class. It is the only way to retrieve the project, since they did not offer a newer support to obtain the project and made it deprecated.

Comment (atothhpe) [reply src/main/java/com/hpe/adm/octane/ideplugins/intellij/ui/detail/GeneralEntityDetailsPanel.java]: @alexandra-marinescu, @kdrtibor  need to test with multiple projects open. In intellij a project is an enitre intellij window, so it's tied to the whole ide frame. We need to see if, with multiple projects/frames open, this code properly detects the project from the focused ide window

File: src/main/java/com/hpe/adm/octane/ideplugins/intellij/ui/detail/FieldsSelectFrame.java (hunk 10 above)
Comment (kdrtibor) [reply src/main/java/com/hpe/adm/octane/ideplugins/intellij/ui/detail/FieldsSelectFrame.java]: Did as asked
